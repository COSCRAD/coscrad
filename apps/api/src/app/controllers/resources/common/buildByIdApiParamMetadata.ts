export default () => ({
    name: 'id',
    required: true,
    example: '2',
});
