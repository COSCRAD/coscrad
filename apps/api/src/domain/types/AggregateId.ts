import { isNonEmptyString } from '@coscrad/validation-constraints';

/**
 * This type alias allows us to change the type of all `Aggregate IDs` in-step
 * in the future.
 */
export type AggregateId = string;

export const isAggregateId = (test: unknown): test is AggregateId => isNonEmptyString(test);
