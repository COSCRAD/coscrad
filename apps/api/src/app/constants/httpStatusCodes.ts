import { ValueType } from '../../lib/types/valueType';

const httpStatusCodes = {
    ok: 200,
    badRequest: 400,
    unauthorized: 401,
    forbidden: 403,
    notFound: 404,
    internalError: 500,
};

export default httpStatusCodes;

export type HttpStatusCode = ValueType<typeof httpStatusCodes>;
