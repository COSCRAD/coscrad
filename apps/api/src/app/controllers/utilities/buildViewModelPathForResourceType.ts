import { ResourceType, resourceTypes } from '../../../domain/types/resourceTypes';

// For irregular plural forms or forms that lead to awkward grammar
const exceptions: Partial<Record<ResourceType, string>> = {
    [resourceTypes.transcribedAudio]: 'transcribedAudioItems',
};

export default (resourceType: ResourceType): string =>
    exceptions[resourceType] || `${resourceType}s`;
