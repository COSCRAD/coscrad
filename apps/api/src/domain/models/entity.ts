import { PartialDTO } from 'apps/api/src/types/partial-dto';
import { EntityType } from '../types/entityType';
import { isNullOrUndefined } from '../utilities/validation/is-null-or-undefined';
import { EntityId } from './types/entity-id';
import { EntityCompositeIdentifier } from './types/entityCompositeIdentifier';

// TODO Extract to a separate location
const isValidStringWithLength = (input: unknown): input is string =>
  typeof input === 'string' && input.length > 0;

/**
 *
 * TODO move this near the corresponding type. Note that this extra layer is
 * meant to safe guard us in case we want stricter validation on an `entity id`.
 * Should we change the rule, we can update this in once place, instead of replacing
 * `isValidStringWithLength` everywhere.
 */
const isValidEntityId = (input: unknown): input is EntityId =>
  isValidStringWithLength(input);

export abstract class Entity {
  readonly id: EntityId;

  abstract readonly type: EntityType;

  readonly published: boolean;

  constructor(dto: unknown) {
    // TODO Validate DTO
    if (!this.#validateDTO(dto))
      throw new Error(`Invalid dto for an entity: ${JSON.stringify(dto)}`);

    this.id = dto.id;

    this.published = typeof dto.published === 'boolean' ? dto.published : false;
  }

  getCompositeIdentifier = (): EntityCompositeIdentifier => ({
    type: this.type,
    id: this.id,
  });

  /**
   * TODO consider moving this to a `Serializable` mixin. There may be cases
   * where we want this behaviour on a non domain-model class without the
   * inheritance baggage.
   *  */
  toDTO<TEntity extends Entity>(this: TEntity): PartialDTO<TEntity> {
    return JSON.parse(JSON.stringify(this));
  }

  /**
   * TODO develop a general pattern \ directory structure for validation \ validators.
   * One idea: Return a symbol Valid | Errors[] from a validator instead of a
   * boolean flag.
   *
   * We may break the validators out into separate files. Then we will validate
   * a DTO before passing it to the constructor in our `instance factory` methods
   * in the mapping layer for the repositories.
   *  */
  #validateDTO = (dto: unknown): dto is PartialDTO<Entity> => {
    const { id } = dto as PartialDTO<Entity>;

    // A new entity will have id auto assigned by db
    if (!isNullOrUndefined(id) && !isValidEntityId(id)) return false;

    return true;
  };
}
