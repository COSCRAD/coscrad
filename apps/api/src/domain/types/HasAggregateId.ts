import { AggregateId } from './AggregateId';

export interface HasAggregateId {
    id: AggregateId;
}
