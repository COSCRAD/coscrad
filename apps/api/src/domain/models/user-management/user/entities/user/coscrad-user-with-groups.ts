import { InternalError } from '../../../../../../lib/errors/InternalError';
import { DTO } from '../../../../../../types/DTO';
import { DeepPartial } from '../../../../../../types/DeepPartial';
import BaseDomainModel from '../../../../BaseDomainModel';
import { CoscradUserGroup } from '../../../group/entities/coscrad-user-group.entity';
import { CoscradUser } from './coscrad-user.entity';

export class CoscradUserWithGroups extends CoscradUser {
    readonly groups: CoscradUserGroup[];

    constructor(user: CoscradUser, userGroups: CoscradUserGroup[]) {
        super(user.toDTO());

        this.groups = userGroups.map((group) => group.clone({}));
    }

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    override clone<T extends BaseDomainModel>(this: T, overrides?: DeepPartial<DTO<T>>): T {
        /**
         * Because this class constructor has a different API, our generic
         * implementation of clone fails.
         *
         * TODO Fix this.
         */
        throw new InternalError(`Not Implemented: CoscradUserWithGroups.clone`);
    }
}
