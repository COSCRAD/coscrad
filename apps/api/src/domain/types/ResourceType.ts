import { UuidDocument } from '../../lib/id-generation/types/UuidDocument';
import { IBibliographicReference } from '../models/bibliographic-reference/interfaces/bibliographic-reference.interface';
import { Book } from '../models/book/entities/book.entity';
import { Category } from '../models/categories/entities/category.entity';
import { EdgeConnection } from '../models/context/edge-connection.entity';
import { MediaItem } from '../models/media-item/entities/media-item.entity';
import { Photograph } from '../models/photograph/entities/photograph.entity';
import { Song } from '../models/song/song.entity';
import { ISpatialFeature } from '../models/spatial-feature/ISpatialFeature';
import { Tag } from '../models/tag/tag.entity';
import { Term } from '../models/term/entities/term.entity';
import { TranscribedAudio } from '../models/transcribed-audio/entities/transcribed-audio.entity';
import { CoscradUserGroup } from '../models/user-management/group/entities/coscrad-user-group.entity';
import { CoscradUser } from '../models/user-management/user/entities/user/coscrad-user.entity';
import { VocabularyList } from '../models/vocabulary-list/entities/vocabulary-list.entity';
import { AggregateId } from './AggregateId';

export enum ResourceType {
    term = 'term',
    vocabularyList = 'vocabularyList',
    transcribedAudio = 'transcribedAudio',
    book = 'book',
    photograph = 'photograph',
    spatialFeature = 'spatialFeature',
    bibliographicReference = 'bibliographicReference',
    song = 'song',
    mediaItem = 'mediaItem',
}

export const isResourceType = (input: unknown): input is ResourceType =>
    Object.values(ResourceType).includes(input as ResourceType);

// We should use this for type inference a few places.
export type ResourceTypeToResourceModel = {
    term: Term;
    vocabularyList: VocabularyList;
    transcribedAudio: TranscribedAudio;
    book: Book;
    photograph: Photograph;
    spatialFeature: ISpatialFeature;
    bibliographicReference: IBibliographicReference;
    song: Song;
    mediaItem: MediaItem;
};

/**
 * This represents the state of all domain models, excluding their `Connections`
 */
export type InMemorySnapshotOfResources = {
    [K in ResourceType]?: ResourceTypeToResourceModel[K][];
};

export type InMemorySnapshot = {
    resources: InMemorySnapshotOfResources;
    connections: EdgeConnection[];
    tags: Tag[];
    /**
     * We do not intend to leak the abstraction of how the categories are
     * represented in the database here. Defer this to (the db specific) document
     * mapping layer.
     */
    categoryTree: Category[];
    users: CoscradUser[];
    userGroups: CoscradUserGroup[];
    uuids: UuidDocument<AggregateId>[];
};
