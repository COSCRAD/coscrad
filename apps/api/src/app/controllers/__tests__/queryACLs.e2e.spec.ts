import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import getValidResourceInstanceForTest from '../../../domain/domainModelValidators/__tests__/domainModelValidators/utilities/getValidResourceInstanceForTest';
import { Resource } from '../../../domain/models/resource.entity';
import { AccessControlList } from '../../../domain/models/shared/access-control/access-control-list.entity';
import { CoscradUserWithGroups } from '../../../domain/models/user-management/user/entities/user/coscrad-user-with-groups';
import { AggregateId } from '../../../domain/types/AggregateId';
import { HasAggregateId } from '../../../domain/types/HasAggregateId';
import { ResourceType } from '../../../domain/types/ResourceType';
import buildInMemorySnapshot from '../../../domain/utilities/buildInMemorySnapshot';
import generateRandomTestDatabaseName from '../../../persistence/repositories/__tests__/generateRandomTestDatabaseName';
import TestRepositoryProvider from '../../../persistence/repositories/__tests__/TestRepositoryProvider';
import buildTestData from '../../../test-data/buildTestData';
import { BaseViewModel } from '../../../view-models/buildViewModelForResource/viewModels/base.view-model';
import formatAggregateType from '../../../view-models/presentation/formatAggregateType';
import httpStatusCodes from '../../constants/httpStatusCodes';
import buildViewModelPathForResourceType from '../utilities/buildViewModelPathForResourceType';
import setUpIntegrationTest from './setUpIntegrationTest';

const { users, userGroups } = buildTestData();

const dummyUser = users[0];

const dummyGroup = userGroups[0].clone({ userIds: [dummyUser.id] });

const dummyUserWithGroups = new CoscradUserWithGroups(dummyUser, [dummyGroup]);

describe('Access Control List and Role Based filtering in resource queries', () => {
    let app: INestApplication;

    // let arangoConnectionProvider: ArangoConnectionProvider;

    let testRepositoryProvider: TestRepositoryProvider;

    beforeAll(async () => {
        ({ app, testRepositoryProvider } = await setUpIntegrationTest(
            {
                ARANGO_DB_NAME: generateRandomTestDatabaseName(),
                BASE_DIGITAL_ASSET_URL: 'https://www.mysound.org/downloads/',
            },
            {
                testUserWithGroups: dummyUserWithGroups,
            }
        ));
    });

    Object.values(ResourceType)
        .filter((resourceType) => resourceType === ResourceType.book)
        .forEach((resourceType) => {
            describe(`when querying for ${formatAggregateType(resourceType)}`, () => {
                const endpointUnderTest = `/${buildViewModelPathForResourceType(resourceType)}`;

                describe('when the user is authenticated, and not a project or COSCRAD admin', () => {
                    beforeEach(async () => {
                        await testRepositoryProvider.testSetup();
                    });

                    afterEach(async () => {
                        // await testRepositoryProvider.testTeardown();
                        // await arangoConnectionProvider.dropDatabaseIfExists();
                        // await app.close();
                    });

                    const unpublishedResourceWithUserInACL = getValidResourceInstanceForTest(
                        resourceType
                    ).clone({
                        published: false,
                        queryAccessControlList: new AccessControlList().allowUser(
                            dummyUserWithGroups.id
                        ),
                    });

                    const numberOfPrivateResourcesToFind = 8;

                    const privateResourcesTheUserCanQuery = Array(numberOfPrivateResourcesToFind)
                        .fill(unpublishedResourceWithUserInACL)
                        .map((resource: Resource, index) =>
                            resource.clone({
                                id: `PRIVATE-ACL-OK-${index.toString()}`,
                            })
                        );

                    const numberOfPublishedResources = 5;

                    const publicResources = Array(numberOfPublishedResources)
                        .fill(unpublishedResourceWithUserInACL.clone({ published: true }))
                        .map((resource: Resource, index) =>
                            resource.clone({
                                id: `PUBLISHED-${(
                                    index + numberOfPrivateResourcesToFind
                                ).toString()}`,
                            })
                        );

                    const confidentialResources = privateResourcesTheUserCanQuery.map(
                        (resource: Resource, index) =>
                            resource.clone({
                                id: `CONFIDENTIAL-${(
                                    index +
                                    numberOfPrivateResourcesToFind +
                                    numberOfPublishedResources
                                ).toString()}`,
                                // The user does not have permission to query this resource
                                queryAccessControlList: new AccessControlList(),
                            })
                    );

                    it('should find unpublished resources when the user is in the ACL', async () => {
                        await testRepositoryProvider.addFullSnapshot(
                            buildInMemorySnapshot({
                                users: [dummyUser],
                                userGroups: [dummyGroup],
                                resources: {
                                    [resourceType]: [
                                        ...privateResourcesTheUserCanQuery,
                                        ...publicResources,
                                        ...confidentialResources,
                                    ],
                                },
                            })
                        );

                        const resourcesThatShouldBeFound = [
                            ...publicResources,
                            ...privateResourcesTheUserCanQuery,
                        ];

                        const res = await request(app.getHttpServer()).get(
                            `/resources${endpointUnderTest}`
                        );

                        expect(res.status).toBe(httpStatusCodes.ok);

                        const viewModels = res.body.data.map((result) => result.data);

                        expect(viewModels.every((viewModel) => viewModel instanceof BaseViewModel));

                        // TODO move this to the functional lib and use it everywhere
                        const getId = ({ id }: HasAggregateId): AggregateId => id;

                        const allFoundViewModelIds = viewModels.map(getId).sort();

                        expect(allFoundViewModelIds).toEqual(
                            resourcesThatShouldBeFound.map(getId).sort()
                        );

                        expect(viewModels).toMatchSnapshot();
                    });
                });
            });
        });
});
