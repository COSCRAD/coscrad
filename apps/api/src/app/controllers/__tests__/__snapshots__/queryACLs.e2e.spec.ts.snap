// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Access Control List and Role Based filtering in resource queries when querying for Book when the user is authenticated, and not a project or COSCRAD admin should find unpublished resources when the user is in the ACL 1`] = `
Array [
  Object {
    "author": "John Doeburg",
    "id": "PRIVATE-ACL-OK-0",
    "publicationDate": "2002",
    "subtitle": "The wolf's revenge",
    "tags": Array [],
    "title": "Three Little Pigs",
  },
  Object {
    "author": "John Doeburg",
    "id": "PRIVATE-ACL-OK-1",
    "publicationDate": "2002",
    "subtitle": "The wolf's revenge",
    "tags": Array [],
    "title": "Three Little Pigs",
  },
  Object {
    "author": "John Doeburg",
    "id": "PRIVATE-ACL-OK-2",
    "publicationDate": "2002",
    "subtitle": "The wolf's revenge",
    "tags": Array [],
    "title": "Three Little Pigs",
  },
  Object {
    "author": "John Doeburg",
    "id": "PRIVATE-ACL-OK-3",
    "publicationDate": "2002",
    "subtitle": "The wolf's revenge",
    "tags": Array [],
    "title": "Three Little Pigs",
  },
  Object {
    "author": "John Doeburg",
    "id": "PRIVATE-ACL-OK-4",
    "publicationDate": "2002",
    "subtitle": "The wolf's revenge",
    "tags": Array [],
    "title": "Three Little Pigs",
  },
  Object {
    "author": "John Doeburg",
    "id": "PRIVATE-ACL-OK-5",
    "publicationDate": "2002",
    "subtitle": "The wolf's revenge",
    "tags": Array [],
    "title": "Three Little Pigs",
  },
  Object {
    "author": "John Doeburg",
    "id": "PRIVATE-ACL-OK-6",
    "publicationDate": "2002",
    "subtitle": "The wolf's revenge",
    "tags": Array [],
    "title": "Three Little Pigs",
  },
  Object {
    "author": "John Doeburg",
    "id": "PRIVATE-ACL-OK-7",
    "publicationDate": "2002",
    "subtitle": "The wolf's revenge",
    "tags": Array [],
    "title": "Three Little Pigs",
  },
  Object {
    "author": "John Doeburg",
    "id": "PUBLISHED-8",
    "publicationDate": "2002",
    "subtitle": "The wolf's revenge",
    "tags": Array [],
    "title": "Three Little Pigs",
  },
  Object {
    "author": "John Doeburg",
    "id": "PUBLISHED-9",
    "publicationDate": "2002",
    "subtitle": "The wolf's revenge",
    "tags": Array [],
    "title": "Three Little Pigs",
  },
  Object {
    "author": "John Doeburg",
    "id": "PUBLISHED-10",
    "publicationDate": "2002",
    "subtitle": "The wolf's revenge",
    "tags": Array [],
    "title": "Three Little Pigs",
  },
  Object {
    "author": "John Doeburg",
    "id": "PUBLISHED-11",
    "publicationDate": "2002",
    "subtitle": "The wolf's revenge",
    "tags": Array [],
    "title": "Three Little Pigs",
  },
  Object {
    "author": "John Doeburg",
    "id": "PUBLISHED-12",
    "publicationDate": "2002",
    "subtitle": "The wolf's revenge",
    "tags": Array [],
    "title": "Three Little Pigs",
  },
]
`;
